### PR TITLE
Installing and Configuring WordPress using WP-CLI

### PR DESCRIPTION
**Pull request by navjotjsingh**
Hey everyone!

This is a guide regarding how to install WordPress and do some basic stuff like managing themes and plugins using WP-CLI.

Let me know how to proceed.

### CODE REVIEW

**Comment by alexfornuto** [comment]
Sorry for the incorrect references, that was a PR # from a different system.

**Comment by navjotjsingh** [comment]
No issues.

**Comment by navjotjsingh** [comment]
Any updates?

File: docs/websites/cms/install-and-configure-wordpress-using-wp-cli.md [hunk 1]
```diff
@@ -0,0 +1,266 @@
...
+
+Or
+
+	wget https://raw.githubusercontent.com/wp-cli/builds/gh-pages/phar/wp-cli.phar
+
+The following command will check whether the downloaded copy is working correctly or not.
+
+	php wp-cli.phar --info
+
+You would get an output mentioning the version of WP-CLI installed and some basic information. That means, we can move ahead.
```
Comment (alexfornuto) [inline docs/websites/cms/install-and-configure-wordpress-using-wp-cli.md]: Specific output please.

File: docs/websites/cms/install-and-configure-wordpress-using-wp-cli.md [hunk 2]
```diff
@@ -0,0 +1,266 @@
...
+### Activating Bash-Completion
+
+Now it is difficult all commands at once and you might want a quick reference on the command you need. While you can always type `wp` and press enter to see the complete list. But wouldn't it be better if you can point to specific command. Like if you want to know what all sub commands are under the `core` parameter. Bash completion script of WP-CLI can help here. To install the script, make sure you are in the home directory of the user you are using.
+
+	cd /home/user
+	wget https://github.com/wp-cli/wp-cli/raw/master/utils/wp-completion.bash
+
+Now edit the .bash_profile so that it is loaded by the shell everytime you login.
+
+	nano ~/.bash_profile
```
Comment (alexfornuto) [inline docs/websites/cms/install-and-configure-wordpress-using-wp-cli.md]: At guides of this level we assume the reader has a preferred text editor and knows how to use it. Instead of showing the command to open the file, just show what needs to be edited with a {: .file-excerpt} section.

File: docs/websites/cms/install-and-configure-wordpress-using-wp-cli.md [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+
+Add the following line in the editor.
+
+	source /<fullpath>/wp-completion.bash
```
Comment (alexfornuto) [inline docs/websites/cms/install-and-configure-wordpress-using-wp-cli.md]: We don't use `<>` to reference variables in code blocks.

File: docs/websites/cms/install-and-configure-wordpress-using-wp-cli.md [hunk 4]
```diff
@@ -0,0 +1,266 @@
...
+	
+`:` is a prompt where you need to enter few commands to navigate through this help menu. Up and down arrow keys will take you through the complete help. And pressing q will exit the help menu. That's all you need to know for now. For complete detail on how to navigate through the complete help section, you can always type h at the above prompt.
+
+In our last step, we enabled Bash Completion feature for WP-CLI. To use that type `wp` and press tab twice. You would see the list of available commands. You can repeat the same by typing 'wp core' and then pressing tab twice. Now you will see a list of commands that can be used with `core`.
+
+## Installing WordPress
+
+### Seting up Database
+
+Before you proceed, you need to setup a database first. For that, login to the MySQL server first.
```
Comment (alexfornuto) [inline docs/websites/cms/install-and-configure-wordpress-using-wp-cli.md]: Whenever there's a variable to change in a code block, announce it in the description above. In this case: Replace `user` with your MySQL user.

File: docs/websites/cms/install-and-configure-wordpress-using-wp-cli.md [hunk 5]
```diff
@@ -0,0 +1,266 @@
...
+
+	grant all on wordpress.* to 'user' identified by 'password';
+
+Type `quit` to exit the MySQL commandline.
+
+First of all, move to the user directory where you would be installing the blog. Traditionally it should be in `/var/www/html` directory. Create a new user directory for your user if it doesn't exist already. Also proper group and ownership permissions need to be given to the user.
+
+	cd /var/wwww/html
+	sudo mkdir wpblog
+	sudo chown -R user:user wpblog
```
Comment (alexfornuto) [inline docs/websites/cms/install-and-configure-wordpress-using-wp-cli.md]: Can you please explain your decision to give `user` permissions over the WordPress directory? 

Also, is `wpblog` a standard example name?

**Comment by alexfornuto** [comment]
In addition to the inline notes above, I've created my own branch on my personal fork with some example formatting changes: https://github.com/alexfornuto/docs/tree/wordpress-cli

Please review both the notes and changes, and take a look at our [Style Guide](https://www.linode.com/docs/style-guide). I look forward to your next update!

**Comment by navjotjsingh** [comment]
I gave user permissions to the WordPress directory as root permissions over that folder are considered insecure. Better to have a non root user have permission over the directory. And no, wpblog is not a standard. Just that I find it convenient whenever I mention it in an example.

I have added all the fixes you suggested. Let me know if there are any issues.

**Comment by navjotjsingh** [comment]
Any updates?

**Comment by alexfornuto** [comment]
Sorry for the delay @navjotjsingh! You are near the top of the queue, and we should be able to review this again with the next week or so.

**Comment by alexfornuto** [comment]
This is the correct tabbing, along with some other formatting fixes: https://github.com/navjotjsingh/docs/pull/1

In addition, any section with multiple steps should use numbering, per our [formatting guide](https://www.linode.com/docs/linode-writers-formatting-guide#paragraph-structure). Please review that entire guide for additional formatting instruction.

**Comment by navjotjsingh** [comment]
Added the numbering and fixed few other errors. Extremely sorry for messing up the formatting.

File: docs/websites/cms/install-and-configure-wordpress-using-wp-cli.md [hunk 6]
```diff
@@ -0,0 +1,273 @@
...
+        WP-CLI root dir:	phar://wp-cli.phar
+        WP-CLI global config:
+        WP-CLI project config:
+        WP-CLI version:	0.21.1
+
+You can use the above procedure for upgrading WP-CLI as well.
+
+### Activating Bash-Completion
+
+Now it is difficult all commands at once and you might want a quick reference on the command you need. While you can always type `wp` and press enter to see the complete list. But wouldn't it be better if you can point to specific command. Like if you want to know what all sub commands are under the `core` parameter. Bash completion script of WP-CLI can help here.
```
Comment (alexfornuto) [inline docs/websites/cms/install-and-configure-wordpress-using-wp-cli.md]: First sentence is incomplete. Third sentence starts off as a question but ends in a period, and could be tightened into a smaller version of the next.

File: docs/websites/cms/install-and-configure-wordpress-using-wp-cli.md [hunk 7]
```diff
@@ -0,0 +1,273 @@
...
+
+1.  To install the script, make sure you are in the home directory of the user(`username` in this case) you are using.
+
+        cd /home/username
+        wget https://github.com/wp-cli/wp-cli/raw/master/utils/wp-completion.bash
+
+2.  Now edit the .bashrc so that it is loaded by the shell every time you login. Open the file and add the following line in the editor assuming you downloaded the file in the home directory.
```
Comment (alexfornuto) [inline docs/websites/cms/install-and-configure-wordpress-using-wp-cli.md]: Because you told the reader explicitly where to download the file, you don't have to assume its location.

File: docs/websites/cms/install-and-configure-wordpress-using-wp-cli.md [hunk 8]
```diff
@@ -0,0 +1,273 @@
...
+        ~~~
+
+3.  Now run the following command to reload the Bash profile.
+
+        source ~/.bashrc
+
+That's it. Bash Completion is enabled.
+
+## Basics of WP-CLI
+Before we get into the helm of the things, let us learn some essential basics of how exactly WP-CLI works. This would help you feel comfortable with what is going to follow next.
```
Comment (alexfornuto) [inline docs/websites/cms/install-and-configure-wordpress-using-wp-cli.md]: I've never before heard or read the term "into the helm of things". Can you rephrase to something last colloquial?

File: docs/websites/cms/install-and-configure-wordpress-using-wp-cli.md [hunk 9]
```diff
@@ -0,0 +1,273 @@
...
+
+3.  Now we need to grant all the required privileges for the database to the mysql user.
+
+        grant all on wordpress.* to 'user' identified by 'password';
+
+4.  Type `quit` to exit the MySQL command line.
+
+### Main Install
+
+1.  Move to the user directory where you would be installing the blog. Traditionally it should be in `/var/www/html` directory. Create a new user directory for your user if it doesn't exist already. Also proper group and ownership permissions need to be given to the user.
```
Comment (alexfornuto) [inline docs/websites/cms/install-and-configure-wordpress-using-wp-cli.md]: This section is confusing because it references both a user directory and `/var/www/html` as the document root. Because you have specified our LAMP guide as a prerequisite, you can safely assume that the reader is using `/var/www/html` and that the user is www-data.

File: docs/websites/cms/install-and-configure-wordpress-using-wp-cli.md [hunk 10]
```diff
@@ -0,0 +1,273 @@
...
+
+        cd /var/www/html
+        sudo mkdir wpblog
+        sudo chown -R user:user wpblog
+        cd wpblog
+
+    {: .note}
+    >
+    >This guide for the sake of simplicity assumes that you will be hosting only one site i.e. the blog on your VPS and therefore doesn't go into the detail of setting up a virtual host file. Should you need to host multiple sites, you will need to refer to our [guide on Apache Virtual hosts](/docs/websites/apache/apache-web-server-on-ubuntu-14-04#configure-name-based-virtual-hosts) for the same.
```
Comment (alexfornuto) [inline docs/websites/cms/install-and-configure-wordpress-using-wp-cli.md]: As noted above, these details are expressed in the prerequisite guide.

File: docs/websites/cms/install-and-configure-wordpress-using-wp-cli.md [hunk 11]
```diff
@@ -0,0 +1,273 @@
...
+
+2.  Next, download the WordPress files.
+
+        wp core download
```
Comment (alexfornuto) [inline docs/websites/cms/install-and-configure-wordpress-using-wp-cli.md]: Running this as the non-root user will require `sudo`

File: docs/websites/cms/install-and-configure-wordpress-using-wp-cli.md [hunk 12]
```diff
@@ -0,0 +1,273 @@
...
+
+3.  Create a wp-config.php file.
+
+        wp core config --dbname=Database --dbuser=user --dbpass=password --dbhost=localhost --dbprefix=wp_
```
Comment (alexfornuto) [inline docs/websites/cms/install-and-configure-wordpress-using-wp-cli.md]: Above you have the reader create the database `wordpress`, so please use that in the example. Also, please explain to the reader all other variables that need to be changed in this and any command.

File: docs/websites/cms/install-and-configure-wordpress-using-wp-cli.md [hunk 13]
```diff
@@ -0,0 +1,273 @@
...
+
+    dbhost and dbprefix are entirely optional and can be omitted unless you need to change their default values.
+
+4.  Run the installation.
+
+        wp core install --url="<yourdomain>" --title="Blog Title" --admin_user="adminusername" --admin_password="password" --admin_email="emailid"
```
Comment (alexfornuto) [inline docs/websites/cms/install-and-configure-wordpress-using-wp-cli.md]: Never use `< >` symbols for variables the reader needs to change, unless they actually need to wrap the variables with those characters.

File: docs/websites/cms/install-and-configure-wordpress-using-wp-cli.md [hunk 14]
```diff
@@ -0,0 +1,273 @@
...
+
+Your WordPress blog is ready for use.
+
+## Common Commands
+
+### Installing and Updating Plugins
+
+Let's say we need to install Yoast SEO plugin. First step is to search for it to find the slug.
```
Comment (alexfornuto) [inline docs/websites/cms/install-and-configure-wordpress-using-wp-cli.md]: I think this would be a good place to explain the term `slug` for readers who haven't seen it before.

File: docs/websites/cms/install-and-configure-wordpress-using-wp-cli.md [hunk 15]
```diff
@@ -0,0 +1,272 @@
...
+
+3.  Now we need to grant all the required privileges for the database to the mysql user.
+
+        grant all on wordpress.* to 'user' identified by 'password';
+
+4.  Type `quit` to exit the MySQL command line.
+
+### Main Install
+
+1.  Move to the Apache document root directory and give appropriate ownership to the public_html directory.
```
Comment (alexfornuto) [inline docs/websites/cms/install-and-configure-wordpress-using-wp-cli.md]: You need an empty line before the code excerpt. Why would you be changing the ownership of the html document root to the user?

**Comment by navjotjsingh** [comment]
Because WP-CLI needs access to the public_html directory to download the files during installation.It would fail otherwise.

**Comment by alexfornuto** [comment]
@navjotjsingh Changing the ownership is a bad idea, it reduces security and increases unreliability. There are several other options, each with their pros and cons:

Add the user to the apache user's group.
Use `sudo -u www-data` for commands within that directory
Elevate to root with `su` or `sudo su` to perform the remaining steps.

Also, please check my fork of this repo at alexfornuto/docs, where I've made another commit condensing and straightening up some of the explanation.

**Comment by navjotjsingh** [comment]
I am sorry but I don't understand. I have some doubts.

You want me to leave the ownership of  'public_html' to root user? Also with what you are suggesting (sudo -u www-data) would still not run a command like 'wp core download' because the folder is still owned by root.

**Comment by navjotjsingh** [comment]
Anyway. I have incorporated all the fixes suggested by you at your repo. Thanks.

**Comment by alexfornuto** [comment]
Ownership of `/var/www/html` should be either root or the user apache runs as. The use one of the methods listed above to interact with it.

**Comment by navjotjsingh** [comment]
Have kept the public_directory with root user as is. Modified wp-cli usage with --allow-root parameter where it was required.

**Comment by alexfornuto** [comment]
Per the wp-cli interface (when you don't add `allow-root`):

```
Error: YIKES! It looks like you're running this as root. You probably meant to run this as the user that your WordPress install exists under.

If you REALLY mean to run this as root, we won't stop you, but just bear in mind that any code on this site will then have full control of your server, making it quite DANGEROUS.

If you'd like to continue as root, please run this again, adding this flag:  --allow-root

If you'd like to run it as the user that this site is under, you can run the following to become the respective user:

    sudo -u USER -i -- wp <command>
```

If you look at our existing [WordPress guide](https://www.linode.com/docs/websites/cms/how-to-install-and-configure-wordpress) you'll see that we have ownership of these files set to the user `www-data`. I know that I suggested either changing to this user or to root, but in light of the dangers of having your WordPress installation owned by the root user, I would strongly prefer that you use the www-data user, either by running your `wp` commands as `www-data`, or by changing ownership after installation.

@felicianotech, you're pretty knowledgable when it comes to installing and configuring WordPress. I'd love any input you may have on this.

**Comment by felicianotech** [comment]
I run all of my website instances (including WordPress) as its own user. Then, when I need to work on it, I log in directly as that user or I `su` to that user.

If I didn't do that, I would have the sites owned by the www-data user and add my personal user to that group.

This way you can use the cli as yourself or the user of the website with no problem.

**Comment by alexfornuto** [comment]
Thanks for the input @felicianotech!

Since our website hosting guides don't set up individual users per domain, using the www-data user us the way to go.

**Comment by navjotjsingh** [comment]
Made the required changes. Please check.

**Comment by navjotjsingh** [comment]
Any updates?

**Comment by navjotjsingh** [comment]
@kcraig34 Accepted your merge request.

**Comment by navjotjsingh** [comment]
Any updates?

**Comment by alexfornuto** [comment]
You bet! Check out https://github.com/navjotjsingh/docs/pull/3 - once you merge that I think we're ready to publish. Please let us know if you prefer a Linode credit or PayPal, and in either case what email address to use.

**Comment by navjotjsingh** [comment]
I just did. And I would prefer a Paypal payment at navjotjsingh (at) gmail (dot) com.

**Comment by navjotjsingh** [comment]
Got the money. Thanks. :)

**Comment by alexfornuto** [comment]
You're welcome! Thanks for your contribution.
